File: samples/python/extensions/timestamp/tests/test_client.py
# ruff: noqa: S101
import datetime
import os
import subprocess
import sys
import time

from pathlib import Path

import httpx
import pytest

from a2a.client import ClientConfig, ClientFactory
from a2a.types import (
    AgentCapabilities,
    AgentCard,
    AgentInterface,
    Message,
    Part,
    Role,
    SendMessageRequest,
    TaskState,
)
from timestamp_ext.client import wrap_client_factory
from timestamp_ext.core import TIMESTAMP_FIELD, TimestampExtension


_AGENT_URL = 'http://127.0.0.1:9998'
_FIXED_TS = 1_700_000_000.0


@pytest.fixture(scope='session', autouse=True)
def start_server():
    server_path = Path(__file__).parent / '__main__.py'
    env = os.environ.copy()
    env['TIMESTAMP_EXT_FIXED_CLOCK'] = str(_FIXED_TS)
    process = subprocess.Popen(  # noqa: S603
        [sys.executable, str(server_path)],
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        env=env,
    )
    # Wait a moment for the server to start
    time.sleep(1.5)

    yield

    process.terminate()
    process.wait()


@pytest.mark.asyncio
async def test_timestamp_extension_round_trip():
    expected_iso = datetime.datetime.fromtimestamp(_FIXED_TS, datetime.timezone.utc).isoformat()
    ext = TimestampExtension(now_fn=lambda: _FIXED_TS)

    # Recreate the card on the client side for discovery
    card = AgentCard(
        name='Echo',
        description='echo agent that demonstrates the timestamp extension',
        version='1.0.0',
        default_input_modes=['text'],
        default_output_modes=['text'],
        capabilities=AgentCapabilities(streaming=True),
        supported_interfaces=[
            AgentInterface(
                protocol_binding='JSONRPC',
                url=_AGENT_URL,
                protocol_version='1.0',
            )
        ],
    )
    # Add capabilities to match the server
    ext.add_to_card(card=card)

    async with httpx.AsyncClient(base_url=_AGENT_URL) as httpx_client:
        factory = wrap_client_factory(
            factory=ClientFactory(config=ClientConfig(httpx_client=httpx_client, streaming=True)),
            ext=ext,
        )
        client = factory.create(card=card)

        request = SendMessageRequest(
            message=Message(
                role=Role.ROLE_USER,
                parts=[Part(text='hi')],
                message_id='req-1',
            )
        )

        print('\n--- streaming response from the agent ---')
        artifacts = []
        status_messages = []
        async for chunk in client.send_message(request=request):
            kind = chunk.WhichOneof('payload')
            if chunk.HasField('artifact_update'):
                art = chunk.artifact_update.artifact
                artifacts.append(art)
                print(f'  artifact "{art.name}" @ {art.metadata[TIMESTAMP_FIELD]}')
            elif chunk.HasField('status_update'):
                status = chunk.status_update.status
                if status.HasField('message'):
                    msg = status.message
                    status_messages.append(msg)
                    print(
                        f'  status={TaskState.Name(status.state)} message '
                        f'@ {msg.metadata[TIMESTAMP_FIELD]}'
                    )
                else:
                    print(f'  status={TaskState.Name(status.state)}')
            else:
                print(f'  event of kind {kind}')

        await client.close()

    assert artifacts, 'agent did not emit an artifact'
    assert status_messages, 'agent did not emit a status message'
    for art in artifacts:
        assert art.metadata[TIMESTAMP_FIELD] == expected_iso
    for msg in status_messages:
        assert msg.metadata[TIMESTAMP_FIELD] == expected_iso

File: samples/js/extensions/timestamp/test_client.ts
/* eslint-disable n/no-missing-import */
import {
    AgentCard,
    Artifact,
    Message,
    Role,
    SendMessageRequest,
    TaskState,
} from "@a2a-js/sdk";
import {
    ClientFactory,
    DefaultAgentCardResolver,
    JsonRpcTransportFactory,
} from "@a2a-js/sdk/client";
import { v4 as uuidv4 } from "uuid";

import {
    wrapClientFactory,
    TIMESTAMP_FIELD,
    TimestampExtension,
} from "./timestamp_ext/index.js";

const AGENT_URL = process.env.AGENT_URL || "http://127.0.0.1:9998";
const TIMESTAMP_UNIX = 1700000000.0;

async function getAgentCard(): Promise<AgentCard> {
    const resolver = new DefaultAgentCardResolver();
    const card = await resolver.resolve(AGENT_URL);
    return card;
}

async function main(): Promise<void> {
    const expectedIso = new Date(TIMESTAMP_UNIX * 1000).toISOString();
    // Initialize with TIMESTAMP_UNIX for validation
    const ext = new TimestampExtension(() => TIMESTAMP_UNIX);

    console.log(`Resolving agent card from ${AGENT_URL}...`);
    const card = await getAgentCard();

    console.log("Creating client factory and client...");
    const baseFactory = new ClientFactory({
        transports: [new JsonRpcTransportFactory()],
    });
    const factory = wrapClientFactory(baseFactory, ext);
    const client = await factory.createFromAgentCard(card);

    const request: SendMessageRequest = {
        tenant: "",
        metadata: {},
        message: {
            messageId: uuidv4(),
            role: Role.ROLE_USER,
            parts: [
                {
                    content: { $case: "text", value: "hi" },
                    mediaType: "text/plain",
                    filename: "",
                    metadata: {},
                },
            ],
            taskId: "",
            contextId: "",
            extensions: [],
            metadata: {},
            referenceTaskIds: [],
        },
        configuration: undefined,
    };

    console.log("\n--- streaming response from the agent ---");
    const stream = client.sendMessageStream(request);

    const artifacts: Artifact[] = [];
    const statusMessages: Message[] = [];

    for await (const chunk of stream) {
        if (chunk.payload?.$case === "artifactUpdate") {
            const art = chunk.payload.value.artifact;
            if (art) {
                artifacts.push(art);
                const timestamp = art.metadata?.[TIMESTAMP_FIELD];
                console.log(`  artifact "${art.name}" @ ${timestamp}`);
            }
        } else if (chunk.payload?.$case === "statusUpdate") {
            const status = chunk.payload.value.status;
            if (status) {
                const stateName = TaskState[status.state];
                if (status.message) {
                    statusMessages.push(status.message);
                    const timestamp =
                        status.message.metadata?.[TIMESTAMP_FIELD];
                    console.log(`  status=${stateName} message @ ${timestamp}`);
                } else {
                    console.log(`  status=${stateName}`);
                }
            }
        } else {
            console.log(`  event of kind ${chunk.payload?.$case}`);
        }
    }

    // Assertions / Verification
    console.log("\nVerifying timestamps...");
    if (artifacts.length === 0) {
        throw new Error("Verification failed: Agent did not emit an artifact");
    }
    if (statusMessages.length === 0) {
        throw new Error(
            "Verification failed: Agent did not emit a status message"
        );
    }

    for (const art of artifacts) {
        const ts = art.metadata?.[TIMESTAMP_FIELD];
        if (ts !== expectedIso) {
            throw new Error(
                `Verification failed: Expected artifact timestamp ${expectedIso}, got ${ts}`
            );
        }
    }
    for (const msg of statusMessages) {
        const ts = msg.metadata?.[TIMESTAMP_FIELD];
        if (ts !== expectedIso) {
            throw new Error(
                `Verification failed: Expected message timestamp ${expectedIso}, got ${ts}`
            );
        }
    }

    console.log("Verification successful! All timestamps match.");
}

main().catch((err) => {
    console.error("Error running test client:", err);
    // eslint-disable-next-line n/no-process-exit
    process.exit(1);
});
